### PR TITLE
server: fix nfs version option during mounts

### LINKED ISSUE #9469: secstorage.nfs.version is not used in two scenarios
the NFS version is not considered in two scenarios

- During systemvm template registration

https://github.com/apache/cloudstack/blob/7214c13e106368127c928b2028c80c3bacaf528c/engine/schema/src/main/java/com/cloud/upgrade/SystemVmTemplateRegistration.java#L87

https://github.com/apache/cloudstack/blob/7214c13e106368127c928b2028c80c3bacaf528c/engine/schema/src/main/java/com/cloud/upgrade/SystemVmTemplateRegistration.java#L436-L437

- When cleanup diagnosis files

https://github.com/apache/cloudstack/blob/7214c13e106368127c928b2028c80c3bacaf528c/server/src/main/java/org/apache/cloudstack/diagnostics/DiagnosticsServiceImpl.java#L479-L481

the nfs version is set to `null`

##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~

~~~

##### CLOUDSTACK VERSION

~~~

~~~

##### STEPS TO REPRODUCE

~~~

~~~

##### EXPECTED RESULTS

~~~

~~~

##### ACTUAL RESULTS

~~~

~~~

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #9469

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

File: engine/schema/src/main/java/com/cloud/upgrade/SystemVmTemplateRegistration.java [hunk 1]
```diff
@@ -888,4 +901,17 @@ public void doInTransactionWithoutResult(final TransactionStatus status) {
             }
         });
     }
+
+    public String getNfsVersion(long storeId) {
+        final String configKey = "secstorage.nfs.version";
```
Comment (weizhouapache) [inline engine/schema/src/main/java/com/cloud/upgrade/SystemVmTemplateRegistration.java:906]: `CapacityManager.ImageStoreNFSVersion.key()` ?

File: engine/schema/src/main/java/com/cloud/upgrade/SystemVmTemplateRegistration.java [hunk 2]
```diff
@@ -326,7 +339,7 @@ public static boolean validateIfSeeded(String url, String path) {
             if (filePath == null) {
                 throw new CloudRuntimeException("Failed to create temporary directory to mount secondary store");
             }
-            mountStore(url, filePath);
+            mountStore(url, filePath, null);
```
Comment (weizhouapache) [inline engine/schema/src/main/java/com/cloud/upgrade/SystemVmTemplateRegistration.java:342]: nfs version is `null` ?

File: engine/schema/src/main/java/com/cloud/upgrade/SystemVmTemplateRegistration.java (hunk 1 above)
Comment (shwstppr) [reply engine/schema/src/main/java/com/cloud/upgrade/SystemVmTemplateRegistration.java:906]: @weizhouapache it doesn't seem to be possible to use ConfigKey in `SystemVmTemplateRegistration` class as it is part of `engine-schema` module. Other configs are accessed the same way. I tried but it creates a cyclical dependency

Comment (weizhouapache) [reply engine/schema/src/main/java/com/cloud/upgrade/SystemVmTemplateRegistration.java:906]: ok, then it is fine to use  "secstorage.nfs.version"
thanks @shwstppr for the attempt

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Review by abh1sar — APPROVED** [review]
LGTM. Didn't test.

**Review by sureshanaparti — APPROVED** [review]
clgtm

**Review by weizhouapache — COMMENTED** [review]
code lgtm

**Review by weizhouapache — APPROVED** [review]

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@JoaoJandre this looks like a good to have for 4.20 (and 4.19) can you concur?

**Comment by JoaoJandre** [comment]
> @JoaoJandre this looks like a good to have for 4.20 (and 4.19) can you concur?

@DaanHoogland yes but it seems that no one tested the PR (yet). If we can get some manual tests on it before the freeze I think it should be in 4.20

**Comment by DaanHoogland** [comment]
> > @JoaoJandre this looks like a good to have for 4.20 (and 4.19) can you concur?
> 
> @DaanHoogland yes but it seems that no one tested the PR (yet). If we can get some manual tests on it before the freeze I think it should be in 4.20

trying to fit it in @JoaoJandre ;)

**Review by rohityadavcloud — APPROVED** [review]

File: engine/schema/src/main/java/com/cloud/upgrade/SystemVmTemplateRegistration.java [hunk 3]
```diff
@@ -141,6 +146,14 @@ public SystemVmTemplateRegistration(String systemVmTemplateVersion) {
...
     }
 
+    public static String getMountCommand(String nfsVersion, String device, String dir) {
+        String cmd = "sudo mount -t nfs";
+        if (StringUtils.isNotBlank(nfsVersion)) {
+            cmd = String.format("%s -o vers=%s", cmd, nfsVersion);
+        }
+        return String.format("%s %s %s", cmd, device, dir);
+    }
+
```
Comment (DaanHoogland) [inline engine/schema/src/main/java/com/cloud/upgrade/SystemVmTemplateRegistration.java:156]: this looks like a good generic method for creating the mount command, but it is in `SystemVmTemplateRegistration`. Does it make sense to move it to a utility and search for other places to use this, @shwstppr ? (some deduplication may be resulting)

**Comment by DaanHoogland** [comment]
> > > @JoaoJandre this looks like a good to have for 4.20 (and 4.19) can you concur?
> > 
> > 
> > @DaanHoogland yes but it seems that no one tested the PR (yet). If we can get some manual tests on it before the freeze I think it should be in 4.20
> 
> trying to fit it in @JoaoJandre ;)

This got a bit behind in priority, need to create an env that doesn't support nfsv3 to do a functional test.

**Comment by DaanHoogland** [comment]
registering a template as system vm works but getting diagnostics data fails so far:
image

image

```
Get diagnostics data
(r-4-VM) Failed to generate diagnostics zip file in VM 4e7941c0-9226-488b-bb8f-70dc2711ab00 due to: java.io.IOException: Stream closed at java.base/java.io.BufferedInputStream.getBufIfOpen(BufferedInputStream.java:176) at java.base/java.io.BufferedInputStream.read(BufferedInputStream.java:342) at java.base/sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:295) at java.base/sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:337) at java.base/sun.nio.cs.StreamDecoder.read(StreamDecoder.java:179) at java.base/java.io.InputStreamReader.read(InputStreamReader.java:181) at java.base/java.io.BufferedReader.fill(BufferedReader.java:161) at java.base/java.io.BufferedReader.readLine(BufferedReader.java:326) at java.base/java.io.BufferedReader.readLine(BufferedReader.java:392) at com.cloud.utils.script.OutputInterpreter.processError(OutputInterpreter.java:38) at com.cloud.utils.script.Script.execute(Script.java:313) at com.cloud.hypervisor.kvm.resource.LibvirtComputingResource.executeInVR(LibvirtComputingResource.java:546) at com.cloud.agent.resource.virtualnetwork.VirtualRoutingResource.execute(VirtualRoutingResource.java:479) at com.cloud.agent.resource.virtualnetwork.VirtualRoutingResource.executeQueryCommand(VirtualRoutingResource.java:222) at com.cloud.agent.resource.virtualnetwork.VirtualRoutingResource.executeRequest(VirtualRoutingResource.java:128) at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtNetworkElementCommandWrapper.execute(LibvirtNetworkElementCommandWrapper.java:35) at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtNetworkElementCommandWrapper.execute(LibvirtNetworkElementCommandWrapper.java:29) at com.cloud.hypervisor.kvm.resource.wrapper.LibvirtRequestWrapper.execute(LibvirtRequestWrapper.java:78) at com.cloud.hypervisor.kvm.resource.LibvirtComputingResource.executeRequest(LibvirtComputingResource.java:1929) at com.cloud.agent.Agent.processRequest(Agent.java:683) at com.cloud.agent.Agent$AgentRequestHandler.doTask(Agent.java:1106) at com.cloud.utils.nio.Task.call(Task.java:83) at com.cloud.utils.nio.Task.call(Task.java:29) at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) at java.base/java.lang.Thread.run(Thread.java:829) 
```

**Pull request merged**